### ISSUE TITLE
Spurious error about `Sized` with string slices and generators

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#45845]
Given [this program](https://play.rust-lang.org/?gist=ed3381e83e0fcba2e3e3954afc6a4978&version=nightly):

```rust
#![feature(generators, generator_trait)]

use std::ops::Generator;

fn main() {
   let s = String::from("foo");
   let mut gen = move || {
       yield &s[..]
   };
   gen.resume();
}
```

the compiler emits:

```
error[E0277]: the trait bound `str: std::marker::Sized` is not satisfied
  --> src/main.rs:10:4
   |
10 |    gen.resume();
   |    ^^^^^^^^^^^^ `str` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `str`
   = note: only the last element of a tuple may have a dynamically sized type

error: aborting due to previous error
```

Originally reported as https://github.com/alexcrichton/futures-await/issues/41

**Comment by Zoxc** [comment]
This appears to be fixed.
